apps/server/src/modules/files/uc/sync-files.uc.ts

```diff
@@ -0,0 +1,75 @@
+/* istanbul ignore file */
+
+import { Injectable } from '@nestjs/common';
+import { Logger } from '@src/core/logger/logger.service';
+import { SyncFilesRepo } from '../repo/sync-files.repo';
+import { SyncContext, SyncOptions, SyncFileItem } from '../types';
+import { SyncFilesMetadataService } from './sync-files-metadata.service';
+import { SyncFilesStorageService } from './sync-files-storage.service';
+
+@Injectable()
+export class SyncFilesUc {
+	constructor(
+		private readonly metadataService: SyncFilesMetadataService,
+		private syncFilesRepo: SyncFilesRepo,
+		private storageService: SyncFilesStorageService,
+		private logger: Logger
+	) {
+		this.logger.setContext(SyncFilesUc.name);
+	}
+
+	async syncFilesForTasks(options: SyncOptions, context: SyncContext): Promise<number> {
+		const itemsToSync: SyncFileItem[] = await this.syncFilesRepo.findTaskFilesToSync(options.aggregationSize);
+		const numFound = itemsToSync.length;
+
+		if (numFound > 0) {
+			await Promise.all(
+				Array.from({ length: options.numParallelPromises }).map(async () => {
+					await this.syncBatch(itemsToSync, context);
+				})
+			);
+		} else {
+			this.logger.log('Not items found. Nothing to do.');
+		}
+
+		return numFound;
+	}
+
+	private async syncBatch(itemsToSync: SyncFileItem[], context: SyncContext): Promise<void> {
+		let item: SyncFileItem | undefined;
+		do {
+			item = itemsToSync.shift();
+			if (item) {
+				// eslint-disable-next-line no-await-in-loop
+				await this.syncFile(item, context);
+			}
+		} while (item);
+	}
+
+	private async syncFile(item: SyncFileItem, context: SyncContext) {
+		try {
+			context.fileCounter += 1;
+			const countInfo = context.fileCounter;
+			const fileInfo1 = `source file with id = ${item.source.id}, size = ${item.source.size}`;
+			this.logger.log(`#${countInfo} Start syncing ${fileInfo1}`);
+
+			await this.metadataService.syncMetaData(item);
```

Comment:
It must be execute step by step in context of throwing errors?
What is happend when an error is throw on last step?